BootstrapNAS

ljaljushkin opened the pull request:
### Changes

Introduce training part of the BootstrapNAS algorithm. 

### Reason for changes

NAS capabilities in NNCF

### Related tickets

50799, 73815

### Tests

unit, sanity and e2e tests. 


The leftovers that will be closed before merge
- [x] fix pylint
- [x] add valid JSON schema
- [x] move to experimental
- [x] validate on bigger scope of tests: 
  - [x] real life ?
  - [x] e2e full (builds 712 and 752) 
  - [x] third-party sanity (builds 370 and 406)
  - [x] sota-eval-pytorch (builds 770 was 825)
  - [x] compression training (builds 408)
- [x] add typehints
- [x] resolve TODO in tests

ljaljushkin commented on nncf/torch/dynamic_graph/wrappers.py:
```diff
@@ -87,65 +89,27 @@ def wrapped(*args, **kwargs):
-                if not ctx._elastic_depth or not ctx.in_skipped_block:
-                    result = operator(*args, **kwargs)
-                else:
+                if ctx.elastic_depth and ctx.in_skipped_block:
```
I've significantly refactored `wrapped_operator` function.
@asenina @vshampor  please pay extra attention to it

ljaljushkin replied:
now it's not calling pre/post ops in the skipping mode.
the logic is the same, reorganized code to reduce number of if-conditions

asenina replied:
LGTM

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -0,0 +1,362 @@
+        if self._mode == ElasticDepthMode.MANUAL:
+            skipped_blocks = self._elasticity_params.get('skipped_blocks', [])  # type: List[List[str,str]]
+            self._skipped_blocks = [BuildingBlock(*b) for b in skipped_blocks]
+            self._skip_dependencies = get_group_of_dependent_blocks(self._skipped_blocks)
```
@asenina @jpablomch skip_dependencies is not needed for Elastic Depth in manual mode anymore.

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -0,0 +1,362 @@
+            for block in self._skipped_blocks:
+                start_node = graph.get_node_by_name(block.start_node_name)
+                end_node = graph.get_node_by_name(block.end_node_name)
+                ordinal_ids.append([start_node.node_id, end_node.node_id])
```
@asenina @jpablomch ordinal_ids are not needed for Elastic Depth in manual mode anymore.

ljaljushkin commented on nncf/common/graph/graph.py:
```diff
@@ -252,6 +252,13 @@ def get_all_nodes(self) -> List[NNCFNode]:
             all_nodes.append(nncf_node)
         return all_nodes
 
+    def get_all_simple_paths(self, start_node_name: NNCFNodeName, end_node_name: NNCFNodeName):
```
@asenina @vshampor introduced a new function to NNCFNetwork that is helpful for NAS in 2 places
without need to access a private `_nx_graph` field

daniil-lyakhov commented on examples/torch/common/models/classification/efficientnet.py:
```diff
@@ -0,0 +1,27 @@
+wrapper(Conv2dStaticSamePadding)
+
+
+def efficient_net(pretrained=True, num_classes=1000, **kwargs):
+    if pretrained:
+        return EfficientNet.from_pretrained(num_classes=num_classes, **kwargs)
+    return EfficientNet.from_name(num_classes=num_classes, **kwargs)
```
What is the difference between this implementation and the implementation from torchvision?

ljaljushkin replied:
efficientnet from torchvision is available since 0.11 version which is not supported now.
Currently, we have torchvision 0.10 only. as soon as we move to the latest torchvision, this version of efficientnet from external repo won't be needed.

daniil-lyakhov commented on nncf/common/pruning/utils.py:
```diff
@@ -291,14 +314,15 @@ def count_flops_and_weights_per_node(graph: NNCFGraph,
-        flops_numpy = 2 * np.prod(node.layer_attributes.kernel_size) * \
-                      num_in_channels * filters_per_channel * np.prod(output_shapes[name])
-        weights_numpy = np.prod(node.layer_attributes.kernel_size) * num_in_channels * filters_per_channel
+        flops_numpy = 2 * np.prod(kernel_size) * num_in_channels * filters_per_channel * np.prod(output_shapes[name])
```
Let's discuss flops computing formula. I decided to check it, and as far as I understood here should be `input_shapes` instead of `output_shapes`

ljaljushkin replied:
The formula looks correct. `input_shape` is possible to use, but it will be more complicated with considering `padding` and `stride`.

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,684 @@
+        self._width_list = []
+
+    @property
+    def width_list(self) -> List[int]:
```
Docstring would be helpful

ljaljushkin replied:
thanks for the comment! I've added
please review

daniil-lyakhov replied:
Now it's clear for me, thanks!

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,684 @@
+                    nncf_logger.warning("Wrong value for multiplier: {}. Skipping ".format(multiplier))
+                    continue
+                w = int(max_out_channels * multiplier)
+                w = w - (w % 8)
```
Is 8 an alignment constant?

ljaljushkin replied:
The method generates all possible width for the layers to sample from. 8 is just a pre-defined step size and it coincides with the alignment constant.

Makes sense to explicitly extract this constant and add more explanation to doc-strings. Will do that. Thanks for the comment!

ljaljushkin replied:
@daniil-lyakhov extracted constant, added more explanation in docstrings
please review

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,684 @@
+
+
+class ElasticWidthConv2DOp(ElasticWidthOp, nn.Module):
+    def __init__(self, max_out_channels, node_name, elastic_width_params: Dict[str, Any] = None):
+        super().__init__(max_width=max_out_channels, node_name=node_name)
+
```
Could you please clarify what is the difference between `ElasticWidthConv2DOp` and `DynamicConvInputOp`?

ljaljushkin replied:
The first is for trimming output channels (elasticity), the latter - for input ones (adaptation to elasticity - dynamic trimming of input channels).

Will add more documentation to explain it. Thanks for noticing!

ljaljushkin replied:
@daniil-lyakhov renamed classes and added more docstrings. please review
```
ElasticWidthConv2DOp -> ElasticOutputWidthConv2DOp 
DynamicConvInputOp -> ElasticInputWidthConv2DOp 
```

daniil-lyakhov replied:
Now it's clear, thanks!

asenina commented on nncf/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -317,21 +336,24 @@ def get_state(self) -> Dict[str, Any]:
+                    node = graph.get_node_by_key(node_key)
+                    all_nodes_in_block.add(node.node_name)
+            start_node = graph.get_node_by_key(start_node_key)
+            start_op_address = start_node.node_name
```
LGTM

ljaljushkin commented:
@openvinotoolkit/nncf_pytorch-maintainers any objections to merge?

jpablomch commented on nncf/torch/nas/bootstrapNAS/elasticity/multi_elasticity_handler.py:
```diff
@@ -112,19 +112,20 @@ def activate_supernet(self) -> None:
+            if handler_id in config:
+                sub_config = config[handler_id]
+                other_active_handlers = dict(filter(lambda pair: pair[0] != handler_id, active_handlers.items()))
+                resolved_config = handler.resolve_conflicts_with_other_elasticities(sub_config, other_active_handlers)
```
LGTM.

jpablomch commented on nncf/torch/nas/bootstrapNAS/training/scheduler.py:
```diff
@@ -151,32 +151,38 @@ def get_state(self) -> Dict[str, Any]:
                     f"stage {progressivity_of_elasticity[high_priority_dim_idx]} violates progressivity of elasticity")
-            for i in range(0, high_priority_dim_idx):
-                if i not in stages_covered and progressivity_of_elasticity[i] in enabled_elasticity_dims:
+            for i in range(low_priority_dim_idx, high_priority_dim_idx):
```
Thanks! LGTM.

ljaljushkin commented on nncf/common/graph/graph.py:
```diff
@@ -252,6 +252,13 @@ def get_all_nodes(self) -> List[NNCFNode]:
             all_nodes.append(nncf_node)
         return all_nodes
 
+    def get_all_simple_paths(self, start_node_name: NNCFNodeName, end_node_name: NNCFNodeName):
```
Thanks for noticing! Fixed

AlexKoff88 commented on examples/experimental/torch/classification/nas_advanced_main.py:
```diff
@@ -0,0 +1,150 @@
+"""
```
I would call it like `architecture_search_example.py` or even `bootstrap_nas.py`.

ljaljushkin replied:
Thanks! I've renamed to bootstrap_nas.py

AlexKoff88 commented on examples/torch/common/models/classification/mobilenet_v2_cifar10.py:
```diff
@@ -0,0 +1,154 @@
+"""
```
Just in case. You can use 3rd party projects for CIFAR-10 instead of having your own implementation. Check this one: https://github.com/chenyaofo/pytorch-cifar-models

```
import torch

model = torch.hub.load("chenyaofo/pytorch-cifar-models", "cifar10_mobilenetv2_x1_0", pretrained=True)
```

ljaljushkin replied:
Actually, we use 3rd party project here: https://github.com/huyvnphan/PyTorch_CIFAR10.
Looks like the suggested approach with `torch.hub` is not applicable for this repository, because weights are downloaded for all models once: https://github.com/huyvnphan/PyTorch_CIFAR10/blob/master/data.py#L21

The results in the paper were achieved with these models and weights, that's why switching to the new suggested  repository will require re-running all experiments with mobilenet and resnet and potentially will require changing of some hyperparameters.

@jpablomch  



https://github.com/huyvnphan/PyTorch_CIFAR10/blob/master/data.py#L21

vshampor commented on nncf/torch/dynamic_graph/wrappers.py:
```diff
@@ -148,22 +129,88 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+def _process_ordinary_op(op_address: 'OperationAddress',
```
`_execute_op` could be a more fitting name.

ljaljushkin replied:
Thanks! Fixed

vshampor commented on nncf/torch/pruning/utils.py:
```diff
@@ -61,3 +67,76 @@ def init_output_masks_in_graph(graph: NNCFGraph, nodes: List):
+            modules_in_shapes[node.node_name] = in_shape[0]
+        else:
+            modules_in_shapes[node.node_name] = in_shape[1:]
+    return modules_in_shapes
```
Any file called `utils.py` should not be extended. Introduce a new entity encapsulating the new functionality, separate the code into an own .py file and write corresponding unit tests.

ljaljushkin replied:
It was part of the filter pruning algorithm. I just moved it to common. I hope it's already covered by unit tests. @daniil-lyakhov could you confirm that? 
https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/pruning/filter_pruning/algo.py#L286

daniil-lyakhov replied:
I agree that utils.py is more appropriate place for this functions. We have test which checks mask propagation in disconnected graphs, but tests of flops counting for disconnected graphs are missing. I've reflected it in my ticket 70128

ljaljushkin replied:
Thanks, @daniil-lyakhov

vshampor reviewed (changes requested):
Great job overall! Still, there are some comments that I would like to be addressed.

ljaljushkin commented on nncf/config/config.py:
```diff
@@ -94,17 +98,11 @@ def get_redefinable_global_param_value_for_algo(self, param_name: str, algo_name
-            raise jsonschema.ValidationError(msg)
+        bootstrap_nas_section = loaded_json.get(BOOTSTRAP_NAS_ALGO_NAME_IN_CONFIG)
+        if bootstrap_nas_section is not None:
+            NNCFConfig._validate_json_section_by_schema(bootstrap_nas_section, BOOTSTRAP_NAS_SCHEMA)
```
Because BootstrapNAS is not a compression algorithm with builder and controller - it's built on top of elasticity and training compression algorithms + search part. It also assumes a different integration as opposed to ordinary compression algorithms. That's why it can be added to `EXPERIMENTAL_REF_VS_ALGO_SCHEMA`.

It can be a part of `ROOT_NNCF_CONFIG_SCHEMA` like AA algo:
```json
"accuracy_aware_training": with_attributes(ACCURACY_AWARE_TRAINING_SCHEMA, ...)
```
But this schema shouldn't include experimental attributes.

ljaljushkin replied:
Yes, it's possible. I've generalized it. Thanks!

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+        prunable_types = [NNCFConv2d.op_func_name]
+        self._cluster_next_nodes = get_cluster_next_nodes(graph, self._pruned_module_groups_info, prunable_types)
+
+        self._propagation_graph = deepcopy(graph)
```
I use it for adding some information to nodes. Wanted to isolate this data. 
```
            end_node = self._propagation_graph.get_node_by_name(end_node_name)
            end_mask = end_node.data['output_mask']
```

vshampor replied:
Please add a corresponding comment inline.

ljaljushkin replied:
Thanks! Fixed

alexsu52 commented on nncf/common/graph/graph.py:
```diff
@@ -252,6 +252,24 @@ def get_all_nodes(self) -> List[NNCFNode]:
             all_nodes.append(nncf_node)
         return all_nodes
 
+    def get_all_simple_paths(self,
```
It looks like this is a separate algorithm as in [NetworkX](https://networkx.org/documentation/networkx-1.9/index.html). What do you think about moving this algorithm into a separate module?

ljaljushkin replied:
This method just a wrapper for method from networkx: `nx.all_simple_paths`
I don't think it makes sense to move it to a separate module, because it heavily relies on fields of `NNCFGraph` - 5 entries of `self.`.

alexsu52 commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/base_handler.py:
```diff
@@ -0,0 +1,248 @@
+    @abstractmethod
+    def get_kwargs_for_flops_counting(self) -> Dict[str, Any]:
+        """
+        Provides arguments for counting flops of the currently activated subnet.
```
Honestly, It's not a clear interface.

ljaljushkin replied:
I can introduce `FlopsCountingArgs` class and refactor this method to fill this struct:
```python
def get_kwargs_for_flops_counting(self, args: FlopsCountingParams) -> FlopsCountingParams:
    args.var1=value
    return args
```
@alexsu52 Would it be more clear?

alexsu52 replied:
It's more clear. As I understand you need a list of layers with attributes to calculate Flops. What do you think about returning some subnet descriptor which includes layers and their parameters?

ljaljushkin replied:
I don't feel comfortable to open internals of Elastic Handlers now. I intentionally encapsulated all elasticity related operations and structures inside them. They don't look simple and easy to use outside, moreover incorrect usage may potentially lead to undefined behavior.
```python
node_name_vs_dynamic_input_width_op_map: Dict[NNCFNodeName, ElasticWidthOp],
pruned_module_groups_info: Clusterization[ElasticWidthInfo](id_fn=lambda x: x.node_name),
...
skipped_blocks: BuildingBlocks,
skip_dependencies: GroupedBlockIDs,
node_names_per_block: Dict[int, List[NNCFNodeName]],
ordinal_ids: OrdinalIDs
```

Instead, I propose to introduce a single simple method for each handler:
```python
kernel_sizes = self.kernel_handler.get_active_kernel_sizes_per_node()
names_of_skipped_nodes = self.depth_handler.get_names_of_skipped_nodes()
input_width_values, output_width_values = self.width_handler.get_active_in_out_width_values()
```
If more information is needed outside of handler, it will be not difficult to provide.

ljaljushkin commented on nncf/common/compression.py:
```diff
@@ -46,6 +49,7 @@ class BaseCompressionAlgorithmController(CompressionAlgorithmController):
 
     BUILDER_STATE = 'builder_state'
     CONTROLLER_STATE = 'ctrl_state'
+    _state_names = BaseControllerStateNames
```
`_state_names` shouldn't differ for different instances of the controller, therefore it should be defined for the class. I just moved the declaration from __init__.

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+        :param max_width: maximum value of width
+        :param min_width: minimum value of width
+        :param max_num_params: maximum number of generated width values.
+        :param width_step: step size to select widths starting from minimal width to maximum one.
```
As far as I understand, in case of `width_multipliers` is `None` list is constructed by iteratively substracting `width_step` from `max_width` param e.g. `[max_width - i]` where i changes in range to satisfy the constraints. For me description below is more appropriate
```suggestion
        :param width_step: step size to select widths starting from maximum width to minimum one.
```

ljaljushkin replied:
Thanks! I've paraphrased it to make it more clear.

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+        :param min_width: minimum value of width
+        :param max_num_params: maximum number of generated width values.
+        :param width_step: step size to select widths starting from minimal width to maximum one.
+        :param width_multipliers: list of scale factors (between 0 and 1) that should be independently multiplied with
+        maximum width to get list of width values.
```
Could you please specify, that 
1) `width_multipliers` is mutually exclusive with `width_step` param
2) Alignment constrain equal to 8 is used on width list generation, so values and number of possible widths could differ from just multiplying in vanila case

ljaljushkin replied:
Thanks! I've specified

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+            end_mask = end_node.data['output_mask']
+
+            start_output_shape = list(set(self._propagation_graph.get_output_shapes_for_node(start_node_name)))
+            end_output_shape = list(set(self._propagation_graph.get_output_shapes_for_node(end_node_name)))
```
Could you please explain, why output shapes are wrapping by set firstly and then by list? Set wrapping could change order of shapes dimensions

ljaljushkin replied:
A good catch! Thanks! I've preserved the order using `OrderedDict.from_keys`

daniil-lyakhov commented on nncf/torch/dynamic_graph/wrappers.py:
```diff
@@ -148,22 +129,88 @@ def wrapped(*args, **kwargs):
+        else:
+            retval = module_call(self, *args, **kwargs)
+
+        if type(self).__name__ in ITERATION_MODULES.registry_dict.keys():
```
```suggestion
        if type(self).__name__ in ITERATION_MODULES.registry_dict:
```

ljaljushkin replied:
Thanks! Fixed

daniil-lyakhov commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_kernel.py:
```diff
@@ -0,0 +1,495 @@
+            self.register_parameter(name, param)
+
+    def generate_kernel_size_list(self, max_kernel_size: kernel_size_type) -> List[kernel_size_type]:
+        """
+        Generates list of available kernel size values.
+
+        :param max_kernel_size: maximum value of kernel size
+        :return: list of kernel size values.
+        """
```
Please specify that kernel size should be odd and that default kernel value step is 2

ljaljushkin replied:
Thanks! I've specified

ljaljushkin commented on nncf/torch/dynamic_graph/wrappers.py:
```diff
@@ -148,22 +129,88 @@ def wrapped(*args, **kwargs):
+    return result
+
+
+def _collect_module_attrs_and_ignored_algorithms(ctx, op_name):
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/search_building_blocks/search_blocks.py:
```diff
@@ -497,52 +555,55 @@ def remove_linear_combination(sorted_building_blocks: List[BuildingBlock]) -> Li
     Distributes all nodes to the same output and input shapes.
 
     param: sgraph: SeacrhGraph of target model
     returns: Dict for input/output shapes, where key - shape,
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/search_building_blocks/search_blocks.py:
```diff
@@ -107,18 +117,65 @@ def __eq__(self, other):
 
 
-class BuildingBlock:
+SHAPE_VS_NODES_MAP = Dict[str, Set[SearchGraphNode]]
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/training/training_algorithm.py:
```diff
@@ -0,0 +1,252 @@
+
+    @staticmethod
+    def _define_best_accuracy(acc1, best_acc1,
+                              compression_stage, best_compression_stage) -> Tuple[float, float]:
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/training/training_algorithm.py:
```diff
@@ -0,0 +1,252 @@
+            val_loader,
+            optimizer,
+            checkpoint_save_dir,
+            tensorboard_writer=None) -> Tuple[NNCFNetwork, ElasticityController]:
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/training/scheduler.py:
```diff
@@ -0,0 +1,198 @@
+            first_stage = low_priority_dim_idx
+
+    @staticmethod
+    def _get_default_params() -> Dict[str, List[Dict]]:
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+     """
+    _state_names = EWBuilderStateNames
+
+    def __init__(self, elasticity_params: Optional[Dict[str, Any]] = None,
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+    """
+
+    def __init__(self, target_model: NNCFNetwork,
+                 filter_importance: Callable,
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py:
```diff
@@ -0,0 +1,891 @@
+from nncf.torch.utils import get_model_device
+
+pruning_group_id = int
+width_type = int
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -0,0 +1,382 @@
+from nncf.torch.dynamic_graph.context import TracingContext
+from nncf.torch.nncf_network import NNCFNetwork
+
+block_id = int
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -0,0 +1,382 @@
+    def __init__(self, target_model: NNCFNetwork,
+                 skipped_blocks: BUILDING_BLOCKS,
+                 skip_dependencies: GROUPED_BLOCK_IDS,
+                 node_names_per_block: Dict[int, List[str]],
```
Thanks! Fixed

ljaljushkin commented on nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_depth.py:
```diff
@@ -0,0 +1,382 @@
+                 skipped_blocks: BUILDING_BLOCKS,
+                 skip_dependencies: GROUPED_BLOCK_IDS,
+                 node_names_per_block: Dict[int, List[str]],
+                 ordinal_ids: ORDINAL_IDS):
```
Thanks! Fixed

ljaljushkin commented on nncf/common/pruning/utils.py:
```diff
@@ -272,17 +287,25 @@ def count_flops_and_weights_per_node(graph: NNCFGraph,
         If not specified, taken from the graph. {node_name: channels_num}
+    :param kernel_sizes: Dictionary of kernel sizes in convolutions.
+        If not specified, taken from the graph. {node_name: kernel_size}
+    :param op_addresses_to_skip: List of operation addresses of layers that should be skipped from calculation
```
Thanks! Fixed

ljaljushkin commented:
@daniil-lyakhov @asenina @AlexKoff88 @vshampor @alexsu52 Really appreciate your effort to review this long PR. Thank you for your comments!

I've addressed all your comments. Here's the commit with corrections: https://github.com/openvinotoolkit/nncf/pull/1079/commits/0ca9dbd961285312df3e34b68becc0f21472e0a6

ljaljushkin commented:
Jenkins please retry a build

ljaljushkin commented:
There are 2 ongoing validations blocking the merge:

- [ ] 3rd-party sanity tests - build 408 WIP

- [ ] real life tests - TBD by validation team

ljaljushkin commented:
In summary, BootstrapNAS feature has successfully passed all checks: 

- real life (build 1478, 1479, 1481 and 1487)
- e2e full (builds 712 and 752)
- third-party sanity (builds 370 and 406)
- sota-eval-pytorch (builds 770 was 825)
- compression training (builds 408)

It received 4 approvals from the stakeholders and no objections/comments from other reviewers. 
I believe that PR is ready for merge.
